transmart-core-db/grails-app/services/org/transmartproject/db/clinical/MultidimensionalDataResourceService.groovy

```diff
@@ -592,69 +647,110 @@ class MultidimensionalDataResourceService implements MultiDimensionalDataResourc
         // check if there are any observations for the concept
-        if (!exists(builder, conceptConstraint)) {
-            throw new InvalidQueryException("No observations found for concept path: ${conceptConstraint.path}")
+        if (!exists(builder, constraint)) {
+            throw new InvalidQueryException("No observations found for query")
         }
 
-        // check if the concept is truly numerical (all textValue are E and all numberValue have a value)
-        // all(A) and all(B) <=> not(any(not A) or any(not B))
-        def valueTypeNotNumericConstraint = new FieldConstraint(
-                operator: Operator.NOT_EQUALS,
+        at.collect {aggregateFieldType(it)}.unique().findAll().each {
+            def wrongValueTypeConstraint = new FieldConstraint(
+                    operator: Operator.NOT_EQUALS,
+                    field: valueTypeField,
+                    value: it,
+            )
+            if(exists(builder, new AndConstraint(args: [(Constraint) constraint, wrongValueTypeConstraint]))) {
+                throw new InvalidQueryException('One of the concepts/observations has the wrong type for this aggregation')
+            }
+        }
+
+        // check if the concept is truly numerical (all textValue are E and all numberValue have a value) or textual
+
+        def textTypeConstraint = new FieldConstraint(
+                operator: Operator.EQUALS,
+                field: valueTypeField,
+                value: ObservationFact.TYPE_TEXT,
+        )
+
+        def numberTypeConstraint = new FieldConstraint(
+                operator: Operator.EQUALS,
                 field: valueTypeField,
-                value: ObservationFact.TYPE_NUMBER
+                value: ObservationFact.TYPE_NUMBER,
         )
-        def textValueNotEConstraint = new FieldConstraint(
-                operator: Operator.NOT_EQUALS,
+
+        def textValueEConstraint = new FieldConstraint(
+                operator: Operator.EQUALS,
                 field: textValueField,
                 value: "E"
         )
-        def numberValueNullConstraint = new NullConstraint(
-                field: numberValueField
-        )
-        def notNumericalCombination = new Combination(
-                operator: Operator.OR,
-                args: [valueTypeNotNumericConstraint, textValueNotEConstraint, numberValueNullConstraint]
-        )
-        def conceptNotNumericalCombination = new Combination(
-                operator: Operator.AND,
-                args: [conceptConstraint, notNumericalCombination]
-        )
 
-        if (exists(builder, conceptNotNumericalCombination)) {
-            def message = 'One of the observationFacts had either an empty numerical value or a ' +
-                    'textValue with something else then \'E\''
-            throw new InvalidQueryException(message)
-        }
+        // Transmart currently does not allow null values in the ObservationFacts, but it could be extended to allow
```

Comment:
Why does it not allow that? Because of the exception that is thrown in `ObservationFact` or is there some deeper reason?